presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -226,4 +227,155 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
+
+        List<Long> addresses = new ArrayList<>();
+        for (int index = 0; index < addressRanges.size(); index += 2) {
+            long minAddress = addressRanges.get(index);
+            long maxAddress = addressRanges.get(index + 1);
+            for (long address = minAddress; address <= maxAddress; address++) {
+                addresses.add(address);
+            }
+        }
+        return addresses;
+    }
+
+    /**
+     * Searches for and outputs ranges of addresses within certain ranges in each dimension.
+     * <p/>
+     * Every two elements in the output list represent the minimum and maximum of an address range.
+     * <p/>
+     * For example, if the list is {1, 5, 10, 14}, the address ranges are from {1, 5} and {10, 14}.
+     *
+     * @param minimumValues the list of minimum values of each dimension
+     * @param maximumValues the list of maximum values of each dimension
+     * @return the list of address ranges within the given value ranges
+     */
+    // TODO: Create own "pair" object
+    public List<Long> zOrderCondensedSearchCurve(List<Integer> minimumValues, List<Integer> maximumValues)
+    {
+        ZOrderCurveSearcher curveSearcher = new ZOrderCurveSearcher(minimumValues, maximumValues);
+        return curveSearcher.findAddresses();
+    }
+
+    /**
+     * Provides a structure to search for addresses within given value ranges.
+     */
+    private class ZOrderCurveSearcher
+    {
+        private final List<Integer> minimumValues;
+        private final List<Integer> maximumValues;
+        private final int[] dimensions;
+        private List<Long> addressRanges;
+
+        /**
+         * Class constructor specifying the value ranges and initializing <code>dimensions</code>.
+         * <p/>
+         * The <code>dimensions</code> array stores how many values are stored at each level of the Z-curve.
+         */
+        public ZOrderCurveSearcher(List<Integer> minimumValues, List<Integer> maximumValues)
+        {
+            this.minimumValues = minimumValues;
+            this.maximumValues = maximumValues;
+
+            // Create mutable version of encodingBits
+            List<Integer> bitLengths = new ArrayList<>(encodingBits);
+            this.dimensions = new int[maxBitLength];
+            for (int dimensionIndex = maxBitLength - 1; dimensionIndex >= 0; dimensionIndex--) {
+                for (int bitLengthIndex = 0; bitLengthIndex < bitLengths.size(); bitLengthIndex++) {
+                    if (bitLengths.get(bitLengthIndex) > 0) {
+                        dimensions[dimensionIndex]++;
+                        bitLengths.set(bitLengthIndex, bitLengths.get(bitLengthIndex) - 1);
+                    }
+                }
+            }
+        }
+
+        public List<Long> findAddresses()
+        {
+            addressRanges = new ArrayList<>();
+            findAddressesRecursively(maxBitLength - 1, 0, totalBitLength);
+            return addressRanges;
```

Comment:
We are always returning an empty `addressRanges` list . Is this required ?

Reply:
`findAddressRecursively` updates the `addressRanges` variable.

Reply:
This is a bit hard to read (initializing `addressRanges` outside of the constructor). We should either initialize it in the constructor, or make it a local variable and pass to the function.